File: dispatch_once单例/dispatch_once单例/ViewController.m
//
//  ViewController.m，
//  dispatch_once单例，这个就是股票震荡大结果吗。真的就是这个结果吗
//
//  Created by 高李军 on 15/11/19
//  Copyright © 2015年 高李军. All rights reserved.
//

#import "ViewController.h"
#import "Userfault.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //GCD所调用的类
    //libdispatch.dylib
    [self GCDDemo7];

}


-(void)GCDDemo7{
    
    
    //GCDDemo6 的原型,停在那里无关紧要，都一样，都是一样的
    
    //创建一个队列 异步
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_CONCURRENT);
    
    dispatch_group_t group = dispatch_group_create();
    
    dispatch_group_enter(group);
    
    dispatch_async(queue, ^{
        //这个就是异步和同步的区别，可以开启线程
        NSLog(@"hahahahaahha===%@",[NSThread currentThread]);
       dispatch_group_leave(group);
    });
    
    dispatch_async(queue, ^{
        NSLog(@"哈哈哈啊哈哈%@",[NSThread currentThread]);
        
    });
    
    dispatch_group_notify(group, queue, ^{
        NSLog(@"执行完毕了==%@",[NSThread currentThread]);
    });
}


//组调度
-(void)GCDDemo6{
    
    dispatch_group_t group = dispatch_group_create();
    
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_CONCURRENT);;
    
    dispatch_group_async(group, queue, ^{
        NSLog(@"第一组hahahahah===%@",[NSThread currentThread]);
    });
    
    dispatch_group_async(group, queue, ^{
        NSLog(@"第二组hahahahahah-----%@",[NSThread currentThread]);
    });
    
    
    dispatch_group_notify(group, queue, ^{
        NSLog(@"任务已经执行完毕");
    });
}






-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    
    //因为是纯C语言的，所以字符串的前边不应该添加@
    //创建一个队列，一个并发队列
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_CONCURRENT);

    //然后将任务添加在队列中,这个才是真正的任务，就是真正的renwu
    
    for (int i = 0; i<10; i++) {
        dispatch_async(queue, ^{
            NSLog(@"打印结果%@===%d",[NSThread currentThread],i);
        });
        
        NSLog(@"come  here");
    }
    
    //原来是这个样子添加任务，真的就是不懂啊，不懂啊，添加任务,在
}

-(void)GCDDemo4{
    
    //队列是并行队列，任务是异步任务,这个会开启多条线程，然后执行顺序也是乱序的
    
    //创建一个并行队列
    dispatch_queue_t queueu = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_CONCURRENT);
    
    //将任务添加啊到这个并行队列中
    for (int i = 0; i<10; i++) {
        dispatch_async(queueu, ^{
            NSLog(@"%@=====%d",[NSThread currentThread],i);
        });
        
        NSLog(@"come here");//竟然先打印出来的come here
    }
}


-(void)GCDDemo3{
    
    //队列是并行队列，然后同步任务
    
    //创建一个队列，是并行的,同步任务不会开启线程，执行顺序是一个接一个，上一个没有执行完毕，下一个绝对不会执行
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_CONCURRENT);
    
    //将同步任务添加到这个异步队列中
    for (int i = 0; i<10; i++) {
        dispatch_sync(queue, ^{
            NSLog(@"%@===%d",[NSThread currentThread],i);
        });
        
        NSLog(@"com here");
    }
    
    
    
}

-(void)GCDDemo2{
    //串行队列，异步任务，会开线程，但是里面的任务还是会一个一个的执行
    //创建一个串行队列
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_SERIAL);
    
    //将异步任务添加到串行队列中
    for (int i = 0; i<10 ; i++) {
        dispatch_async(queue, ^{
            NSLog(@"%@=======%d",[NSThread currentThread], i);
        });
        
        NSLog(@"come here %@",[NSThread currentThread]);
    }
}




//重新整理一哈GCD的基本用法
-(void)GCDDemo1{
    //同步任务，串行队列，串行队列在创建的时候使用到的一个枚举的值DISPATCH_QUEUE_SERIAL
    
    //创建一个并发队列
    dispatch_queue_t queue = dispatch_queue_create("gaolijun", DISPATCH_QUEUE_SERIAL);
    
    //将任务添加到队列里面，将任务添加
    
    for (int i = 0; i< 10; i++) {
        dispatch_sync(queue, ^{
            NSLog(@"%@===%d",[NSThread currentThread],i);
        });
        NSLog(@"come here");
    }
    //总结：串行队列，同步任务；不开启线程，就是在当前线程执行，然后再执行任务的时间按照一定的顺序执行
    //真的
}




- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: dispatch_group_enter的练习/dispatch_group_enter的练习/ViewController.swift
//
//  ViewController.swift
//  dispatch_group_enter的练习
//
//  Created by 高李军 on 15/10/31.
//  Copyright © 2015年 DamaiPlayBusinessPhone. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    
        //建立一个 dispatch_group,可以减痛一组异步任务完成以后，统一得到通知
        let group = dispatch_group_create()
        
        //进入群组
        dispatch_group_enter(group)
        
        
        //离开群组
        dispatch_group_leave(group)
        
        
        //监听群组调度
        dispatch_group_notify(group, dispatch_get_main_queue()) { () -> Void in
            //执行回调
        }
        
        
        //创建一个group
        let  group1 = dispatch_group_create()
        
        
        dispatch_group_enter(group1)
        
        dispatch_group_leave(group1)
        
        dispatch_group_notify(group1, dispatch_get_main_queue()) { () -> Void in
            //执行回调
        }

        /*
        // 3. 监听群组调度
        dispatch_group_notify(group, dispatch_get_main_queue(), { () -> Void in
        println("图片缓存完成+加载数据 \(NSHomeDirectory())")
        // 执行回调
        completion(statuses: statuses)
        
        
        */
    
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

/*
// 建立一个 dispatch_group，可以监听一组异步任务完成后，统一得到通知
let group = dispatch_group_create()

// 遍历微博数据数组
for s in statuses as [Status]! {
// 没有图片继续循环
if s.pictureURLs == nil {
continue
}

// 遍历图片 url 数组，下载图像
for url in s.pictureURLs! {
// downloadImageWithURL 本身会下载图片，并且完成缓存
// 回调参数，通常是有 UI 交互才会使用的

// 1> 进入群组
dispatch_group_enter(group)
// SDWebImage 下载图像前，会检查本地是否存在图片缓存，如果存在，会直接返回，不会去网络上加载
// 速度快，效率高！
// 如果要判断服务器的图片是否更改，可以使用 RefreshCached 选项
SDWebImageManager.sharedManager().downloadImageWithURL(url, options: nil, progress: nil, completed: { (_, _, _, _, _) in

// 2> 离开群组，一定要放在 block 的最后一句，表示异步真正执行完成
dispatch_group_leave(group)
})
}
}

// 3. 监听群组调度
dispatch_group_notify(group, dispatch_get_main_queue(), { () -> Void in
println("图片缓存完成+加载数据 \(NSHomeDirectory())")
// 执行回调
completion(statuses: statuses)
})
*/


//OC版的dispatch_group_enter的异步执行
/*
#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
[super viewDidLoad];
// Do any additional setup after loading the view, typically from a nib.
}

/**
$ man dispatch_group_enter

void
dispatch_group_async(dispatch_group_t group, dispatch_queue_t queue, dispat
ch_block_t block)
{
dispatch_retain(group);
dispatch_group_enter(group);
dispatch_async(queue, ^{
block();
dispatch_group_leave(group);
dispatch_release(group);
});
}
*/
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {


// 1. 创建一个群组


dispatch_group_t group1 = dispatch_group_create();



dispatch_group_t group = dispatch_group_create();

dispatch_queue_t q = dispatch_get_global_queue(0, 0);

// 2. 异步任务 － 需要注意 enter & leave 成对出现！
// 群组准备监听后续异步方法的执行，打了一个标记
dispatch_group_enter(group);
dispatch_async(q, ^{
[NSThread sleepForTimeInterval:0.5];
NSLog(@"1.--%@", [NSThread currentThread]);

// block 的最后一句，宣告任务完成，取消在群组中的标记
dispatch_group_leave(group);
});

dispatch_group_enter(group);
dispatch_async(q, ^{
[NSThread sleepForTimeInterval:0.5];
NSLog(@"2.--%@", [NSThread currentThread]);

// block 的最后一句，宣告任务完成，取消在群组中的标记
dispatch_group_leave(group);
});

// 3. 监听任务执行，当 group 中所有标记都没有的时候，得到通知
dispatch_group_notify(group, dispatch_get_main_queue(), ^{
NSLog(@"完成回调 %@", [NSThread currentThread]);
});
}

- (void)groupDemo {
// 1. 创建一个群组
dispatch_group_t group = dispatch_group_create();

dispatch_queue_t q = dispatch_get_global_queue(0, 0);

// 2. 异步任务
dispatch_group_async(group, q, ^{
[NSThread sleepForTimeInterval:0.5];
NSLog(@"1.--%@", [NSThread currentThread]);
});
dispatch_group_async(group, q, ^{
[NSThread sleepForTimeInterval:0.8];
NSLog(@"2.--%@", [NSThread currentThread]);
});

// 3. 监听任务执行
dispatch_group_notify(group, dispatch_get_main_queue(), ^{
NSLog(@"完成回调 %@", [NSThread currentThread]);
});
}

@end

*/







